### PR TITLE
SUS-812 | recentchanges - do not use rc_user_text (various places)

### PR DESCRIPTION
**Pull request by macbre**
https://wikia-inc.atlassian.net/browse/SUS-812

### `ApiQueryWatchlist`

Tested on:

* http://poznan.wikia.com/api.php?action=query&list=watchlist&wllimit=50&wlprop=ids|title|timestamp|user|comment&wlexcludeuser=Pyrabot
* http://poznan.wikia.com/api.php?action=query&list=watchlist&wllimit=50&wlprop=ids|title|timestamp|user|comment

```sql
Query plpoznan (DB user: wikia_mw2) (10) (slave): SELECT /* ApiQueryWatchlist::run Macbre - e14e0d14-a110-4c2b-adb0-53f84762af3e */  rc_namespace,rc_title,rc_timestamp,rc_type,rc_cur_id,rc_this_oldid,rc_last_oldid,rc_user,rc_comment  FROM `recentchanges` INNER JOIN `watchlist` ON (wl_user = '119245' AND (wl_namespace=rc_namespace) AND (wl_title=rc_title)) LEFT JOIN `page` ON ((rc_cur_id=page_id))  WHERE rc_deleted = '0' AND (rc_this_oldid=page_latest) AND (rc_timestamp > '')  ORDER BY rc_timestamp DESC LIMIT 51  

Query plpoznan (DB user: wikia_mw2) (11) (slave): SELECT /* ApiQueryWatchlist::run Macbre - c0941dd9-b365-43d3-861e-7ac87df8c3b4 */  rc_namespace,rc_title,rc_timestamp,rc_type,rc_cur_id,rc_this_oldid,rc_last_oldid,rc_user,rc_comment  FROM `recentchanges` INNER JOIN `watchlist` ON (wl_user = '119245' AND (wl_namespace=rc_namespace) AND (wl_title=rc_title)) LEFT JOIN `page` ON ((rc_cur_id=page_id))  WHERE rc_deleted = '0' AND (rc_this_oldid=page_latest) AND (rc_user != 5059646) AND (rc_timestamp > '')  ORDER BY rc_timestamp DESC LIMIT 51
```

### `ApiQueryRecentChanges`

* http://poznan.macbre.wikia-dev.pl/api.php?action=query&list=recentchanges&rcprop=user|userid|comment|title|timestamp&rclimit=200&rcuser=Pyrabot

```sql
Query plpoznan (DB user: wikia_mw2) (11) (slave): SELECT /* ApiQueryRecentChanges::run Macbre - dbb37f44-cbc9-4ef2-8507-b3e03e0284b3 */  rc_timestamp,rc_namespace,rc_title,rc_cur_id,rc_type,rc_moved_to_ns,rc_moved_to_title,rc_deleted,rc_comment,rc_user,rc_user_text  FROM `recentchanges` FORCE INDEX (rc_user)   WHERE rc_deleted = '0' AND rc_user = '5059646'  ORDER BY rc_timestamp DESC LIMIT 201  
```

### CODE REVIEW

**Review by mszabo-wikia — APPROVED** [review]

File: includes/api/ApiQueryWatchlist.php [hunk 1]
```diff
@@ -270,7 +269,7 @@ private function extractRowInfo( $row ) {
 		if ( $this->fld_user || $this->fld_userid ) {
 
 			if ( $this->fld_user ) {
-				$vals['user'] = $row->rc_user_text;
+				$vals['user'] = User::whoIs( $row->rc_user );
```
Comment (mszabo-wikia) [inline includes/api/ApiQueryWatchlist.php]: Should we use `User::getUsername` here for fallback?

**Review by mszabo-wikia — CHANGES_REQUESTED** [review]

File: includes/api/ApiQueryWatchlist.php (hunk 1 above)
Comment (macbre) [reply includes/api/ApiQueryWatchlist.php]: `rc_user_text` column has been removed from the SELECT query - https://github.com/Wikia/app/pull/13879/commits/14581ca643cc5e2c578b5c322892082f9feb0fcd#diff-9dfbf29bb88cc62b4c488c34e2af9aaaL103

Comment (macbre) [reply includes/api/ApiQueryWatchlist.php]: Good point, added in e60a5f3

File: includes/api/ApiQueryRecentChanges.php [hunk 2]
```diff
@@ -188,15 +188,29 @@ public function run( $resultPageSet = null ) {
...
 		if ( !is_null( $params['user'] ) ) {
-			$this->addWhereFld( 'rc_user_text', $params['user'] );
-			$index['recentchanges'] = 'rc_user_text';
+			// SUS-812: handle anon cases (IP address provided) and account names (user name provided)
+			if ( IP::isIPAddress( $params['user'] ) ) {
+				$this->addWhereFld( 'rc_user_text', $params['user'] );
+				$index['recentchanges'] = 'rc_user_text';
+			}
+			else {
+				$this->addWhereFld( 'rc_user', User::newFromName( $params['user'] )->getId() );
```
Comment (mszabo-wikia) [inline includes/api/ApiQueryRecentChanges.php]: Let' use `User::idFromName` helper here

File: includes/api/ApiQueryRecentChanges.php [hunk 3]
```diff
@@ -188,15 +188,29 @@ public function run( $resultPageSet = null ) {
...
 			// * it would require us to sort by rc_user_text before rc_timestamp
 			// * the != condition doesn't throw out too many rows anyway
-			$this->addWhere( 'rc_user_text != ' . $this->getDB()->addQuotes( $params['excludeuser'] ) );
+
+			// SUS-812: handle anon cases (IP address provided) and account names (user name provided)
+			if ( IP::isIPAddress( $params['excludeuser'] ) ) {
+				$this->addWhere( 'rc_user_text != ' . $this->getDB()->addQuotes( $params['excludeuser'] ) );
+			}
+			else {
+				$this->addWhere( 'rc_user != ' . User::newFromName( $params['excludeuser'] )->getId() );
```
Comment (mszabo-wikia) [inline includes/api/ApiQueryRecentChanges.php]: Per above

**Review by mszabo-wikia — CHANGES_REQUESTED** [review]

File: includes/api/ApiQueryRecentChanges.php (hunk 3 above)
Comment (macbre) [reply includes/api/ApiQueryRecentChanges.php]: 0a4d850

File: extensions/AbuseFilter/Views/AbuseFilterViewExamine.php [hunk 4]
```diff
@@ -216,7 +223,15 @@ function __construct( $page, $changesList ) {
 
 	function getQueryInfo() {
 		$dbr = wfGetDB( DB_SLAVE );
-		$conds = array( 'rc_user_text' => $this->mPage->mSearchUser );
+
+		// SUS-812: handle anon cases (IP address provided) and account names (user name provided)
+		if ( IP::isIPAddress( $this->mPage->mSearchUser ) ) {
+			$conds = array( 'rc_user_text' => $this->mPage->mSearchUser );
+		}
+		else {
```
Comment (mszabo-wikia) [inline extensions/AbuseFilter/Views/AbuseFilterViewExamine.php:231]: Indent

File: extensions/AbuseFilter/Views/AbuseFilterViewExamine.php [hunk 5]
```diff
@@ -216,7 +223,15 @@ function __construct( $page, $changesList ) {
 
 	function getQueryInfo() {
 		$dbr = wfGetDB( DB_SLAVE );
-		$conds = array( 'rc_user_text' => $this->mPage->mSearchUser );
+
+		// SUS-812: handle anon cases (IP address provided) and account names (user name provided)
+		if ( IP::isIPAddress( $this->mPage->mSearchUser ) ) {
+			$conds = array( 'rc_user_text' => $this->mPage->mSearchUser );
```
Comment (mszabo-wikia) [inline extensions/AbuseFilter/Views/AbuseFilterViewExamine.php:229]: We must make sure that querying for IP address with any notation works properly - let's pass it through `IP::sanitizeIP`

File: extensions/Nuke/Nuke_body.php [hunk 6]
```diff
@@ -218,8 +218,15 @@ protected function getNewPages( $username, $limit ) {
 
 		if ( $username === '' ) {
 			$what[] = 'rc_user_text';
+			$what[] = 'rc_user'; // SUS-812
 		} else {
-			$where['rc_user_text'] = $username;
+			// SUS-812: handle anon cases (IP address provided) and account names (user name provided)
+			if ( IP::isIPAddress( $username ) ) {
+				$where['rc_user_text'] = $username;
```
Comment (mszabo-wikia) [inline extensions/Nuke/Nuke_body.php:225]: `IP::sanitizeIP` +indent

File: includes/api/ApiQueryRecentChanges.php [hunk 7]
```diff
@@ -188,22 +188,30 @@ public function run( $resultPageSet = null ) {
 		}
 
 		if ( !is_null( $params['user'] ) ) {
-			$this->addWhereFld( 'rc_user_text', $params['user'] );
-			$index['recentchanges'] = 'rc_user_text';
+			// SUS-812: handle anon cases (IP address provided) and account names (user name provided)
+			if ( IP::isIPAddress( $params['user'] ) ) {
+				$this->addWhereFld( 'rc_user_text', $params['user'] );
+				$index['recentchanges'] = 'rc_user_text';
```
Comment (mszabo-wikia) [inline includes/api/ApiQueryRecentChanges.php:194]: `IP::sanitizeIP` + indent 😉

File: includes/api/ApiQueryRecentChanges.php [hunk 8]
```diff
@@ -188,22 +188,30 @@ public function run( $resultPageSet = null ) {
...
-		}
 
-		//  *** Wikia change (@author ADi) *** /Begin
-		if(!is_null($params['user'])) {
-			$this->addWhereFld('rc_user_text', $this->prepareUsername( $params['user'] ));
+			// SUS-812: handle anon cases (IP address provided) and account names (user name provided)
+			if ( IP::isIPAddress( $params['excludeuser'] ) ) {
+				$this->addWhere( 'rc_user_text != ' . $this->getDB()->addQuotes( $params['excludeuser'] ) );
+			}
+			else {
```
Comment (mszabo-wikia) [inline includes/api/ApiQueryRecentChanges.php:211]: `IP::sanitizeIP` + indent 😜

**Review by mszabo-wikia — CHANGES_REQUESTED** [review]

File: includes/revisiondelete/RevisionDeleteUser.php [hunk 9]
```diff
@@ -95,6 +95,15 @@ private static function setUsernameBitfields( $name, $userId, $op, $dbw ) {
 			array( 'rc_user_text' => $name ),
 			__METHOD__
 		);
+
+		// SUS-812: make sure do delete rows for a given user account (by rc_user)
+		$dbw->update(
+			'recentchanges',
+			array( "rc_deleted = rc_deleted $op $delUser" ),
+			array( 'rc_user' => User::idFromName( $name ) ),
```
Comment (mszabo-wikia) [inline includes/revisiondelete/RevisionDeleteUser.php]: We can use `$userId` parameter here.

**Review by mszabo-wikia — CHANGES_REQUESTED** [review]

File: includes/revisiondelete/RevisionDeleteUser.php (hunk 9 above)
Comment (macbre) [reply includes/revisiondelete/RevisionDeleteUser.php]: :+1:  f856180

File: extensions/AbuseFilter/Views/AbuseFilterViewExamine.php (hunk 5 above)
Comment (macbre) [reply extensions/AbuseFilter/Views/AbuseFilterViewExamine.php:229]: As discussed offline, let's keep the code as it is right now and handle IP normalization (via `IP::sanitizeIP`) as a part of IPv6 migration.

**Review by mszabo-wikia — APPROVED** [review]

**Review by ghost — APPROVED** [review]

**Pull request merged**